### ISSUE TITLE
[🐛 Bug]: Selenium/geckodriver/Firefox not working on RHEL (Python)

### ISSUE THREAD

**Issue by gf-salomao** [SeleniumHQ/selenium#12487]
### What happened?

(sorry if this is not the right place to post)
webdriver.Firefox( ) command is not working using Selenium, geckodriver and Firefox on Red Hat Enterprise Linux release 8.5 (Ootpa). I searched and this error is usually a version problem, but versions are compatible following [this link](https://firefox-source-docs.mozilla.org/testing/geckodriver/Support.html)

OS: Red Hat Enterprise Linux release 8.5 (Ootpa)
Selenium version: 4.8.0
geckodriver version: 0.33.0
Firefox version: 102.13.0esr

Python version: 3.8.16

### How can we reproduce the issue?

```shell
By simply using:

>>> from selenium import webdriver
>>> driver = webdriver.Firefox()
```

### Relevant log output

```shell
>>> from selenium import webdriver
>>> driver = webdriver.Firefox()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/firefox/webdriver.py", line 196, in __init__
    super().__init__(command_executor=executor, options=options, keep_alive=True)
  File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 286, in __init__
    self.start_session(capabilities, browser_profile)
  File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 378, in start_session
    response = self.execute(Command.NEW_SESSION, parameters)
  File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 440, in execute
    self.error_handler.check_response(response)
  File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/remote/errorhandler.py", line 245, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: Process unexpectedly closed with status 1
```

### Operating System

Red Hat Enterprise Linux release 8.5 (Ootpa)

### Selenium version

4.8.0

### What are the browser(s) and version(s) where you see this issue?

Mozilla Firefox 102.13.0esr

### What are the browser driver(s) and version(s) where you see this issue?

geckodriver 0.33.0

### Are you using Selenium Grid?

no

**Comment by titusfortner** [comment]
Can you update to Selenium 4.11
Lots of changes and we can't investigate old versions.

**Comment by gf-salomao** [comment]
> Can you update to Selenium 4.11 Lots of changes and we can't investigate old versions.

Just update to version 4.11.2:

```
[root@remote ~]# pip3.8 show selenium
Name: selenium
Version: 4.11.2
Summary: None
Home-page: https://www.selenium.dev
Author: None
Author-email: None
License: Apache 2.0
Location: /usr/local/lib/python3.8/site-packages
Requires: trio-websocket, trio, certifi, urllib3
Required-by: Energy-Lib
[root@remote ~]# python3.8
Python 3.8.16 (default, May 31 2023, 12:44:21) 
[GCC 8.5.0 20210514 (Red Hat 8.5.0-18)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from selenium import webdriver
>>> driver = webdriver.Firefox()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/firefox/webdriver.py", line 67, in __init__
    super().__init__(command_executor=executor, options=options)
  File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 206, in __init__
    self.start_session(capabilities)
  File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 290, in start_session
    response = self.execute(Command.NEW_SESSION, caps)["value"]
  File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 345, in execute
    self.error_handler.check_response(response)
  File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/remote/errorhandler.py", line 229, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: Process unexpectedly closed with status 1
```

**Comment by titusfortner** [comment]
Now can you please turn on logging and send us all the details?
https://www.selenium.dev/documentation/webdriver/troubleshooting/logging/

**Comment by gf-salomao** [comment]
> 

It seems to be a problem on initializing geckodriver?

Here is the log:

```
Executing process: /usr/local/lib/python3.8/site-packages/selenium/webdriver/common/linux/selenium-manager --browser firefox --debug --output json
Checking geckodriver in PATH
Running command: geckodriver --version
Output: "geckodriver 0.33.0 (a80e5fd61076 2023-04-02 18:31 +0000)\n\nThe source code of this program is available from\ntesting/geckodriver in https://hg.mozilla.org/mozilla-central.\n\nThis program is subject to the terms of the Mozilla Public License 2.0.\nYou can obtain a copy of the license at https://mozilla.org/MPL/2.0/."
Running command: which geckodriver
Output: "/usr/local/bin/geckodriver"
Found geckodriver 0.33.0 in PATH: /usr/local/bin/geckodriver
firefox detected at /usr/bin/firefox
Using shell command to find out firefox version
Running command: /usr/bin/firefox -v
Output: "Mozilla Firefox 102.13.0esr"
Detected browser: firefox 102.13.0
Required driver: geckodriver 0.33.0
Driver path: /usr/local/bin/geckodriver
Browser path: /usr/bin/firefox
Using driver at: /usr/local/bin/geckodriver
Started executable: `/usr/local/bin/geckodriver` in a child process with pid: 2279253
```

**Comment by titusfortner** [comment]
What comes after this? It's failing for `driver = webdriver.Firefox()` so we should see what the command to/from the driver is.

**Comment by x86Gr** [comment]
Using
OS Debian 12.1
Python 3.11.2
Firefox 102.14.0 ESR
Geckodriver 0.33.0
Selenium 4.10.0 works fine, but from 4.11.0 onwards it breaks something and the log shows

"Exception trying to discover geckodriver version: error sending request for url (https://github.com/mozilla/geckodriver/releases/latest): error trying to connect: dns error: failed to lookup address information: Try again

selenium and geckodriver are installed in a venv

**Comment by titusfortner** [comment]
@x86Gr are you behind a proxy or otherwise can not access the internet?
Was Selenium downloading the driver for you in 4.10, or do you have it on the computer in a directory on PATH?

**Comment by x86Gr** [comment]
> @x86Gr are you behind a proxy or otherwise can not access the internet? Was Selenium downloading the driver for you in 4.10, or do you have it on the computer in a directory on PATH?

The system has a vanilla connection to the internet and everything resolves and downloads. Selenium used geckodriver on path, i think i've been using them since Debian 11.6 with python 3.9.2 geckodriver 0.32.0 and selenium 4.6.0, never had an issue both online and offline. Basically the system opens Firefox in kiosk mode and receives status data from another process, then displays the current status page using marionette commands

**Comment by titusfortner** [comment]
Selenium 4.6+ manages downloads of drivers, so it needs to look online for updates. 
The change in Selenium 4.11 is that it also looks on PATH for drivers.

@bonigarcia if the driver is on PATH and there is no internet for some reason, it should still locate the driver on PATH? Is this the same issue as what you fixed for IE?

**Comment by bonigarcia** [comment]
> if the driver is on PATH and there is no internet for some reason, it should still locate the driver on PATH?

Yes, the latest versions of Selenium Manager always return the driver PATH (if available) as `result.driver_path`, and, in that case, it is not throwing errors, even without an Internet connection. Here is an example of the execution when the driver is in the PATH and with no Internet connection (notice there is a `WARN` but not an `ERROR`):

```
> selenium-manager --browser firefox --debug

DEBUG   Found geckodriver 0.32.0 in PATH: C:\Users\boni\Documents\bat\geckodriver.exe
DEBUG   firefox detected at C:\Program Files\Mozilla Firefox\firefox.exe
DEBUG   Running command: wmic datafile where name='C:\\Program Files\\Mozilla Firefox\\firefox.exe' get Version /value
DEBUG   Output: "\r\r\n\r\r\nVersion=117.0.0.8615\r\r\n\r\r\n\r\r\n\r"
DEBUG   Detected browser: firefox 117.0.0.8615
WARN    Exception trying to discover geckodriver version: error sending request for url (https://github.com/mozilla/geckodriver/releases/latest): error trying to connect: dns error: No such host is known. (os error 11001)
INFO    Driver path: C:\Users\boni\Documents\bat\geckodriver.exe
INFO    Browser path: C:\Program Files\Mozilla Firefox\firefox.exe
```

> Is this the same issue as what you fixed for IE?

No, that was related to the local metadata.

**Comment by x86Gr** [comment]
> Selenium 4.6+ manages downloads of drivers, so it needs to look online for updates. The change in Selenium 4.11 is that it also looks on PATH for drivers.
> 
> @bonigarcia if the driver is on PATH and there is no internet for some reason, it should still locate the driver on PATH? Is this the same issue as what you fixed for IE?

I can confirm geckodriver is on PATH

**Comment by titusfortner** [comment]
@x86Gr turn on logging so we can see why it isn't detecting it - https://www.selenium.dev/documentation/webdriver/troubleshooting/logging/

**Comment by gf-salomao** [comment]
> What comes after this? It's failing for `driver = webdriver.Firefox()` so we should see what the command to/from the driver is.

Nothing, actually. I removed geckodriver from path to use the selenium version, but the error remains the same. It logs that a executable has started in a child proccess but a WebdriverException happens.

Prompt:
```
# python3.8 selenium_log.py 
Traceback (most recent call last):
  File "selenium_log.py", line 29, in <module>
    main()
  File "selenium_log.py", line 25, in main
    driver = webdriver.Firefox()
  File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/firefox/webdriver.py", line 67, in __init__
    super().__init__(command_executor=executor, options=options)
  File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 206, in __init__
    self.start_session(capabilities)
  File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 290, in start_session
    response = self.execute(Command.NEW_SESSION, caps)["value"]
  File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 345, in execute
    self.error_handler.check_response(response)
  File "/usr/local/lib/python3.8/site-packages/selenium/webdriver/remote/errorhandler.py", line 229, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: Process unexpectedly closed with status 1
```

Log file:
```
# cat selenium_log.txt 
this is useful information
this is a warning
this is detailed debug information
Selenium Manager binary found at: /usr/local/lib/python3.8/site-packages/selenium/webdriver/common/linux/selenium-manager
Executing process: /usr/local/lib/python3.8/site-packages/selenium/webdriver/common/linux/selenium-manager --browser firefox --debug --output json
Checking geckodriver in PATH
Running command: geckodriver --version
Output: ""
geckodriver not found in PATH
firefox detected at /usr/bin/firefox
Using shell command to find out firefox version
Running command: /usr/bin/firefox -v
Output: "Mozilla Firefox 102.13.0esr"
Detected browser: firefox 102.13.0
Required driver: geckodriver 0.33.0
Driver URL: https://github.com/mozilla/geckodriver/releases/download/v0.33.0/geckodriver-v0.33.0-linux64.tar.gz
Driver path: /root/.cache/selenium/geckodriver/linux64/0.33.0/geckodriver
Browser path: /usr/bin/firefox
Using driver at: /root/.cache/selenium/geckodriver/linux64/0.33.0/geckodriver
Started executable: `/root/.cache/selenium/geckodriver/linux64/0.33.0/geckodriver` in a child process with pid: 2958621
```

**Comment by gf-salomao** [comment]
I have successfully resolved the bug.

The RHEL machine I was utilizing is a remote system  with no display. This resulted in the geckodriver failing to launch unless the "headless" option was set within FirefoxOptions(). The error message was not straightforward; however, I deduced the issue by examining the internal log of the geckodriver. The log indicated a failure to locate the DISPLAY environment variable, with a subsequent error a few lines later. Although I was aware that I did not have a display, I configured the headless option on a second instance of the WebDriver but forgot to do on the first one what made me not think about this issue on the first place (sorry).

While I understand that this scenario may be uncommon, I suggest considering the implementation of a feature that provides a more explicit error message within Selenium on this cases (of not having a display), to facilitate troubleshooting in similar situations on the future.

Thanks for the support!
